blockdag/ghostdag.go

```diff
@@ -0,0 +1,156 @@
+		anticoneHeap.Push(parent)
+		queue = append(queue, parent)
+	}
+	for len(queue) > 0 {
+		var current *blockNode
+		current, queue = queue[0], queue[1:]
+		for _, parent := range current.parents {
+			if anticoneSet.contains(parent) || selectedParentPast.contains(parent) {
+				continue
+			}
+			isAncestorOfSelectedParent, err := dag.isAncestorOf(parent, node.selectedParent)
+			if err != nil {
+				return nil, err
+			}
+			if isAncestorOfSelectedParent {
+				selectedParentPast.add(parent)
+				continue
+			}
+			anticoneSet.add(parent)
+			anticoneHeap.Push(parent)
+			queue = append(queue, parent)
+		}
+	}
+	return &anticoneHeap, nil
+}
+
+// blueAnticoneSize returns the blue anticone size of 'block' from the worldview of 'context'.
+// Expects 'block' to be ∈ blue-set(context)
+func (dag *BlockDAG) blueAnticoneSize(block, context *blockNode) (uint32, error) {
+	for current := context; current != nil; current = current.selectedParent {
+		if blueAnticoneSize, ok := current.bluesAnticoneSizes[*block.hash]; ok {
+			return blueAnticoneSize, nil
+		}
+	}
+	return 0, errors.Errorf("block %s is not in blue-set of %s", block.hash, context.hash)
+}
+
+// ghostdag updates newNode.blues, newNode.selectedParent
+// and newNode.bluesAnticoneSizes according to the GHOSTDAG
+// protocol.
+// It updates newNode.blues by going over the anticone of
+// newNode.selectedParent (which is the parent with the
+// highest blue score) and adds it to newNode.blues if it
+// passes two conditions:
+// 1) |anticone(block) ∩ blueset(newNode)| <= K
+// 2) For every blue in blueset(newNode) |anticone(blue) ∩ blueset(newNode) ∪ {block}| <= K.
+//    We do this by maintaining for each block a map bluesAnticoneSizes which holds
+//    all the blue anticone sizes that were affected by the new added blues.
+//    So to find out what is |anticone(blue) ∩ blueset(newNode)| we just iterate in
+//    the selected parent chain of newNode until we find an existing entry in
+//    bluesAnticoneSizes.
+//
+// For further details see the article https://eprint.iacr.org/2018/104.pdf
+func (dag *BlockDAG) ghostdag(newNode *blockNode) (selectedParentAnticone []*blockNode, err error) {
+	newNode.selectedParent = newNode.parents.bluest()
+	newNode.bluesAnticoneSizes[*newNode.hash] = 0
+	newNode.blues = append(newNode.blues, newNode.selectedParent)
+	selectedParentAnticoneHeap, err := dag.selectedParentAnticone(newNode)
+	if err != nil {
+		return nil, err
+	}
+
+	selectedParentAnticone = make([]*blockNode, selectedParentAnticoneHeap.Len())
+	for i := 0; selectedParentAnticoneHeap.Len() > 0; i++ {
+		blueCandidate := selectedParentAnticoneHeap.pop()
+		selectedParentAnticone[i] = blueCandidate
+		candidateBluesAnticoneSizes := make(map[*blockNode]uint32)
+		var candidateAnticoneSize uint32
+		possiblyBlue := true
+
+		for chainBlock := newNode; possiblyBlue; chainBlock = chainBlock.selectedParent {
+			// If blueCandidate is in the future of chainBlock, it means
```

Comment:
I still don't have enough context to understand this comment. By the time that I reach this place in the code I'm completely lost.

I think what I'm missing in the top comment is an overlay of the procedure happening here (all there is right now is the objective). 

I'm not sure the top comment would be the right place for the procedure though, might be better to just add some more comments inside the code, explaining what the code is doing, and why is it doing it.

When writing the comment try to imagine you are someone who has never heard Michael's lecture about ghostdag.